docs/design/object-locking.md

```diff
@@ -0,0 +1,187 @@
+# Object Locking using S3 Object Lock
+
+Object locking feature is built for addressing use cases where
+Write Once Read Many (WORM) model is required. The feature will be built in line
+with AWS specification, any extensions to the specification will be explicitly
+documented. The feature implementation has the goal of meeting
+SEC 17a-4 compliance.
+
+## Requirements
+
+* Bucket needs to have object lock flag set during bucket creation
+* Versioning has to be enabled on the bucket
```

Comment:
not sure if this is the place for it, but we should mention that if the object lock flag is set, versioning will automatically be enabled on the bucket on creation